### ISSUE TITLE
Gradient checkpointing for TF keras models

### ISSUE THREAD

**Issue by pidajay** [tensorflow/tensorflow#38766]
**System information**
- TensorFlow version (you are using): 2.1
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
I have implemented a version of gradient check pointing for TF keras sequential models (with future plans to extend it for Keras functional API and custom models). The PR can be found here - https://github.com/tensorflow/addons/pull/1600. I initially envisioned it as a package in TF addons repo but reviewers felt it was not right place and could potentially go in as a fix for the existing recompute_grad functionality in TF core - https://github.com/tensorflow/tensorflow/blob/64f4a59d5e39b60d67047d5e0b82de0cbcc6c2df/tensorflow/python/ops/custom_gradient.py#L458

Here are the issues with the existing implementation of recompute_grad in TF core and my solutions to those
1. Issue - Not usable. Most people have no idea how to use it. No docs or tutorials that explains how to use it.
Solution - My PR provides a notebook tutorial to demonstrate how to use the implemented functionality.
2. For people who did figure out how to use it, no memory savings was observed.
Solution - My PR provides links to results with observed memory savings. Caveat - only CPU profiled results available. GPU and TPU results need to be done.
3. There is probably an expectation that the user explicitly has to partition the model and decorate each partition. This is not user friendly and can make tasks such as transfer learning difficult.
Solution - My PR expects no explicit partitioning of the model. The user just needs to add a single decorator to the model. That is it.
4. No checkpointing functionality is implemented.
Solution - My PR implements the checkpointing functionality that allows the user to balance the tradeoff between memory and compute time.

Does it make sense to port the PR to TF core?

**Will this change the current api? How?**
The existing implementation can potentially be shoe horned into the existing API for recompute_grad if desirable.

**Who will benefit with this feature?**
Anyone who wants to train models in resource constrained environments.

**Any Other info.**

**Comment by pidajay** [comment]
Does it make sense to port the above mentioned PR as a PR to TF core? Also tagging @seanpmorgan  from TF addons to keep in loop. Thanks!

**Comment by jarednielsen** [comment]
This functionality would be very helpful for training memory-intensive language models!

**Comment by pidajay** [comment]
By the way I am thinking of addressing this with separate PRs as follows
1. Provide a fix (which I already have) and tutorial for the existing tf recompute_grad function. From what I understand, using this requires the user to manually partition the model which might be suboptimal in many cases.
2. Provide a new decorator function '@recompute_sequential' that enables gradient checkpointing on Keras sequential models (like in the PR I have linked above). The big advantage over 1 is that the user does not have to manually partition the model.
3. Provide another decorator function '@recompute_functional' that enables gradient checkpointing on Keras functions models and potentially sub-classed models. 

Does this approach make sense? Thanks!

**Comment by nyngwang** [comment]
Did you make it?

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
